src/jit/stacklevelsetter.cpp

```diff
@@ -0,0 +1,266 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#include "jitpch.h"
+#ifdef _MSC_VER
+#pragma hdrstop
+#endif
+
+#ifndef LEGACY_BACKEND // This file is ONLY used for the RyuJIT backend that uses the linear scan register allocator
+
+#include "stacklevelsetter.h"
+
+StackLevelSetter::StackLevelSetter(Compiler* compiler)
+    : Phase(compiler, "StackLevelSetter", PHASE_STACK_LEVEL_SETTER)
+    , currentStackLevel(0)
+    , maxStackLevel(0)
+#if !FEATURE_FIXED_OUT_ARGS
+    , framePointerRequired(compiler->codeGen->isFramePointerRequired())
+#endif // !FEATURE_FIXED_OUT_ARGS
+    , memAllocator(compiler, CMK_fgArgInfoPtrArr)
+    , putArgNumSlots(&memAllocator)
+{
+}
+
+//------------------------------------------------------------------------
+// DoPhase: Calculate stack slots numbers for outgoing args.
+//
+// Notes:
+//   For non-x86 platforms it calculates the max number of slots
+//   that calls inside this method can push on the stack.
+//   This value is used for sanity checks in the emitter.
+//
+//   For x86 it also sets throw-helper blocks incoming stack depth and set
+//   set framePointerRequired when it is necessary. These values are used to pop
+//   pushed args when an exception occurs.
+void StackLevelSetter::DoPhase()
+{
+    for (BasicBlock* block = comp->fgFirstBB; block != nullptr; block = block->bbNext)
+    {
+        ProcessBlock(block);
+    }
+#if !FEATURE_FIXED_OUT_ARGS
+
+    if (framePointerRequired && !comp->codeGen->isFramePointerRequired())
+    {
+        JITDUMP("framePointerRequired is not set when it is required\n");
+        comp->codeGen->resetWritePhaseForFramePointerRequired();
+        comp->codeGen->setFramePointerRequired(true);
+    }
+#endif // !FEATURE_FIXED_OUT_ARGS
+    assert(maxStackLevel <= comp->fgGetPtrArgCntMax());
+    if (maxStackLevel != comp->fgGetPtrArgCntMax())
+    {
+        JITDUMP("fgPtrArgCntMax was calculated wrong during the morph, the old value: %u, the right value: %u.\n",
```

Comment:
`compiler->fgPtrArgCntMax` is the only value that is used for all arch from the list. It is used only for `noway_assert` in codegencommon.cpp:
```
unsigned maxAllowedStackDepth = compiler->fgPtrArgCntMax +    // Max number of pointer-sized stack arguments.
                                compiler->compHndBBtabCount + // Return address for locally-called finallys
                                genTypeStSz(TYP_LONG) +       // longs/doubles may be transferred via stack, etc
                                (compiler->compTailCallUsed ? 4 : 0); // CORINFO_HELP_TAILCALL args
noway_assert(getEmitter()->emitMaxStackDepth <= maxAllowedStackDepth);
```

if we decide that we can replace it with a regular assert, that we will be able to disable the new phase for release builds on all platforms except x86.
Example where we have this value wrong from System.Private.CoreLib.dll:
```
System.Threading.Interlocked:Add(byref,int):int
fgMorphTree BB34, stmt 17 (before)
[000328] --C-G-------              *  CALL      void   System.Diagnostics.Debug.Assert
[000321] ------------ arg0         +--*  LCL_VAR   int    V19 tmp10
[000324] x---G------- arg1         +--*  IND       ref
[000323] #-----------              |  \--*  IND       int
[000322] ------------              |     \--*  CNS_INT(h) int    0x7389E94 pstr
[000327] x---G------- arg2         \--*  IND       ref
[000326] #-----------                 \--*  IND       int
[000325] ------------                    \--*  CNS_INT(h) int    0x7389E94 pstr
Upping fgPtrArgCntMax from 0 to 1

```

But then we do:
`Removing statement [000332] in BB09 as useless:`
So fgPtrArgCntMax becomes 0. There are also many cases of this error when we sort args of nested calls.